File: translation-script/flatten.py
import typing


def flatten_dict(
    data: dict[str, dict], 
    prevNodes: typing.Optional[list[str]] = None, 
    flattened: typing.Optional[dict[str, str]] = None,
):
    if prevNodes is None:
        prevNodes = []
    if flattened is None:
        flattened = {}

    for key in data.keys():
        child = data.get(key)
        nodes = prevNodes.copy()
        nodes.append(key)
        if isinstance(child, str):
            flattened[".".join(nodes)] = child
        else:
            flatten_dict(child, nodes, flattened)    # type: ignore
            
    return flattened


                
def dict_from_nodes(prev_dict: dict, nodes: list[str], index: int, value: str):
    node = nodes[index]

    if index == len(nodes) - 1:
        prev_dict[node] = value
        return

    if prev_dict.get(node) is None:
        prev_dict[node] = {}
        
    dict_from_nodes(prev_dict[node], nodes, index + 1, value)


def inflate_dict(flattened_dict: dict[str, str]) -> dict:
    inflated_dict = {}
    for key, value in flattened_dict.items():
        nodes = key.split('.')
        dict_from_nodes(inflated_dict, nodes, 0, value)
    return inflated_dict

File: translation-script/flatten.dart
import 'dart:math';

import 'package:linum/widgets/settings_screen/language_selector.dart';

class LanguageEntry {
  final String value;
  LanguageEntry(this.value);

  factory LanguageEntry.fromDynamic(dynamic value) {
    if (value is String) {
      return LanguageEntry(value);
    } else if (value is List) {
      return LanguageEntry(value[0] as String);
    } else {
      print("This should not happen");
      return LanguageEntry("");
    }
  }
}

Map<String, LanguageEntry> flattenMap(
    Map<String, dynamic> map,
    List<String> prevNodes,
    Map<String, LanguageEntry> flattened,
) {
  for (final entry in map.entries) {
    final nodes = List<String>.from(prevNodes)..add(entry.key);
    if (entry.value is List || entry.value is String) {
      flattened[nodes.join(".")] = LanguageEntry.fromDynamic(entry.value);
    } else if(entry.value is Map<String, dynamic>) {
      flattenMap(entry.value as Map<String, dynamic>, nodes, flattened);
    }
  }
  return flattened;
}

void mapFromNodes(Map<String, dynamic> prevMap, List<String> nodes, int index, String value) {
  final node = nodes[index];
  if (index == nodes.length - 1) {
    prevMap[node] = value;
    return;
  }

  if (prevMap[node] == null) {
    prevMap[node] = <String, dynamic>{};
  }

  mapFromNodes(prevMap[node] as Map<String, dynamic>, nodes, index + 1, value);

}

Map<String, dynamic> inflateMap(Map<String, LanguageEntry> flattened) {
  final inflated = <String, dynamic>{};
  for (final entry in flattened.entries) {
    final nodes = entry.key.split('.');
    mapFromNodes(inflated, nodes, 0, entry.value.value);
  }
  return inflated;
}
